### ISSUE TITLE
Pattern unifier does not eta-contract modulo singleton types

### ISSUE THREAD

**Issue by andreasabel** [agda/agda#2556]
```agda
works : let
    X : (⊤ → A) → A
    X = _
  in (x : ⊤ → A) → X (λ y → x y) ≡ x _
works x = refl

test : let
    X : (⊤ → A) → A
    X = _
  in (x : ⊤ → A) → X (λ y → x _) ≡ x _
test x = refl
```
In `test`, the meta is yellow.  Agda does not eta-contract `λ y → x record{}` to `x`.
To implement this, we would need to check whether the argument to `x` is, if not `y`, of a singleton type.  Potentially a performance hazard, but the theory asks for this to work.

**Comment by nad** [comment]
Now I'm only speculating, but another approach might be to eta-expand (maybe lazily). However, I'm guessing that the pattern fragment, as implemented, does not include eta-expanded terms. If that is the case, could one change the implementation so that eta-expanded terms are included in the pattern fragment?

**Comment by andreasabel** [comment]
Yes, but in the end you want to eta-contract to a variable you can abstract over.

I noticed that a fix for this issue would need a type-aware traversal of terms.  Or we implement a special eta-contration procedure that only cares about whether in the end a variable is reached.

Here is an example where the singleton is deeper in the term:
```agda
postulate
  A : Set
  a : A

test2 : let
    X : ((⊤ → A) → A) → A
    X = _
  in (x : (⊤ → A) → A) → X (λ y → x λ (z : ⊤) → y _) ≡ x (λ _ → a)
test2 x = refl
```

**Comment by andreasabel** [comment]
Eta-contraction of record values also does not consider singletons:
```agda
prod-works : let
    X : (A × ⊤ → A × ⊤) → A
    X = _
  in (x : A × ⊤ → A × ⊤) → X (λ y → x (proj₁ y , proj₂ y)) ≡ proj₁ (x (a , _))
prod-works x = refl

prod-test : let
    X : (A × ⊤ → A × ⊤) → A
    X = _
  in (x : A × ⊤ → A × ⊤) → X (λ y → x (proj₁ y , _)) ≡ proj₁ (x (a , _))
prod-test x = refl
```

**Comment by andreasabel** [comment]
Agda could also prune zero-information arguments (however, here we should not take irrelevance into account):
```agda
no-info : let
    X : (A → ⊤) → A
    X = _
  in X _ ≡ a
no-info = refl
```
Maybe going too far, but one could also take the isomorphism `(⊤ → A) ~ A` into account.
```agda
iso : let
    X : (⊤ → A) → A
    X = _
  in (x : A) → X (λ _ → x) ≡ x
iso x = refl
```

**Comment by andreasabel** [comment]
I am working on this, using CheckInternal for a type-directed eta-contraction.  However, this has uncovered sleeping bugs, for instance in record definitions (see #2500).  Currently, progress is blocked here.

**Comment by UlfNorell** [comment]
Can you be more precise (do you remember) what exactly is blocking here? The record part of #2500 was just our old friend *not-a-valid-let-declaration* (#434), so I don't see how you could be blocked on that. Is the problem implicit lambda insertion (#1079/#2099)?

**Comment by jespercockx** [comment]
I guess the thing blocking this issue is the fact that `checkInternal` currently exposes many internal/invisible bugs.
